sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -123,43 +141,142 @@ public void close() {
+    // testing here. Maybe handleData should just throw all exceptions, and catch them
+    // and call onError in the calling code?
+    private void handleData(BeamFnApi.Elements.Data data) {
+      try {
+        LogicalEndpoint key = LogicalEndpoint.of(data.getInstructionReference(), data.getTarget());
+        CompletableFuture<Consumer<BeamFnApi.Elements.Data>> consumer = receiverFuture(key);
+        if (!consumer.isDone()) {
+          LOG.debug(
+              "Received data for key {} without consumer ready. "
+                  + "Waiting for consumer to be registered.",
+              key);
+        }
+        consumer.get().accept(data);
+        if (isFinalizationData(data)) {
+          consumers.remove(key);
+        }
+        /*
+         * TODO: On failure we should fail any bundles that were impacted eagerly
+         * instead of relying on the Runner harness to do all the failure handling.
+         */
+      } catch (ExecutionException | InterruptedException e) {
+        LOG.error(
+            "Client interrupted during handling of data for instruction {} and target {}",
+            data.getInstructionReference(),
+            data.getTarget(),
+            e);
+        outboundObserver.onError(e);
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Client failed to handle data for instruction {} and target {}",
+            data.getInstructionReference(),
+            data.getTarget(),
+            e);
+        outboundObserver.onError(e);
+      }
+    }
+
+    private boolean IsQueueingEnabled() {
+      return instructionIdToQueue != null;
+    }
+
+
+    // TODO please help reveiw the concurrency aspect here. Questions:
+    // - Safe use of ConcurrentHashMap?
+    // - Unnecessary use of ConcurrentHashMap?
+    // - Is there some code that can run concurrently now, which was not before?
     @Override
     public void onNext(BeamFnApi.Elements value) {
       for (BeamFnApi.Elements.Data data : value.getDataList()) {
         try {
-          LogicalEndpoint key =
-              LogicalEndpoint.of(data.getInstructionReference(), data.getTarget());
-          CompletableFuture<Consumer<BeamFnApi.Elements.Data>> consumer = receiverFuture(key);
-          if (!consumer.isDone()) {
-            LOG.debug(
-                "Received data for key {} without consumer ready. "
-                    + "Waiting for consumer to be registered.",
-                key);
-          }
-          consumer.get().accept(data);
-          if (data.getData().isEmpty()) {
-            consumers.remove(key);
+          if (IsQueueingEnabled()) {
+            // TODO Is there a cleaner way do to this?
+            // I was trying to avoid always doing multiple lookups, and always creating
+            // a new instance on the heap
+            SynchronousQueue<Data> queue = instructionIdToQueue.get(data.getInstructionReference());
```

Comment:
Use [Map#computeIfAbsent(key, function)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-). You can find some example usages in our codebase like:
https://github.com/apache/beam/search?q=computeIfAbsent&unscoped_q=computeIfAbsent

It is also threadsafe in the case of multiple threads when used with a concurrenthashmap.

Reply:
Obsolete